### PR TITLE
--match-contract and --match-test changed to glob pattern

### PR DESCRIPTION
**Pull request by jatkz**
Motivation

Closes [#1230](https://github.com/foundry-rs/foundry/issues/1230)

Solution

Patterns take a comma separated string which converts to a vector of globs. Jaro Winkler got lowered a little because glob symbols mess with prefix.

edit: I rebased a few times so its cleaner when it gets reviewed the first time, i did some refactoring.

### CODE REVIEW

File: cli/tests/it/test_cmd.rs [hunk 1]
```diff
@@ -71,8 +70,8 @@ contract Dummy {}
         .unwrap();
 
     // set up command
-    cmd.args(["test", "--match-test", "testA.*", "--no-match-test", "testB.*"]);
-    cmd.args(["--match-contract", "TestC.*", "--no-match-contract", "TestD.*"]);
+    cmd.args(["test", "--match-test", "*testA*", "--no-match-test", "*testB*"]);
```
Comment (mattsse) [inline cli/tests/it/test_cmd.rs:73]: is the leading `*` now necessary?

**Review by mattsse — COMMENTED** [review]
thanks for taking this on.

love the test.

I have some concerns about the leading `*`.

Is this now required when trying to match parts of the name?

This would be breaking behavior.

could you provide some examples how match will change with this PR?

**Comment by jatokz** [comment]
> thanks for taking this on.
> 
> love the test.
> 
> I have some concerns about the leading `*`.
> 
> Is this now required when trying to match parts of the name?

Ohh, you are right. It is a silly mistake on my part.

I tested without the leading `*` and it works fine. I can go through and fix everything this evening.

Thank you for your review

**Comment by mattsse** [comment]
@mds1 wdyt?

I think this is an overall improvement but is technically breaking, right? since no more regex. however this is more in line with how other tools do test matching via cli

**Comment by gakonst** [comment]
I believe people use some regex syntax to exclude stuff, do we still support this here?

**Comment by mds1** [comment]
Will look soon! @jatokz Do you mind leaving a comment here with some examples of what works / what doesn't work for each matcher (contract, test, path), compared to what did/didn't work before? 

From reading the conversation here my main concern would be removing regex support, as I do think that's useful (and have found it useful myself). With good naming conventions + regex it's easy to target only certain test functions during development, so I want to better understand the tradeoffs of this PR

**Comment by jatokz** [comment]
> I believe people use some regex syntax to exclude stuff, do we still support this here?

in this pr, the exclude patterns are also changed to comma separated globs

**Comment by gakonst** [comment]
Thanks king - appreciate the effort

**Comment by jatokz** [comment]
hope everyone had a good holiday

@mds1 I came up with a few examples:

1) Case Sensitive:
testComputeCreate2Address
test_createAdress

glob => *[cC]reate*
re => .*[cC]reate.*  OR (?i)(.*create.*)

2) Match where contains Set or Update
testSetFoo
testUpdateBar

glob =>  *Set*,*Update*  
re => .*(Set|Update).*

3) Contain both words in any order

glob => doesnt seem possible with this lib so would have to do a comma sep list of every combination
(?=.*Foo)(?=.*Bar).*

4) Both words in order

glob => *Foo*Bar*
regex => .*Foo.*Bar.*

there are some other globs patterns available here: https://docs.rs/glob/latest/glob/struct.Pattern.html

**Comment by mds1** [comment]
Thanks for those comparisons, that was helpful! 

Seems simple enough of a change UX-wise. In general I think I'm indifferent on globs vs. regex, and since globs can accomplish pretty much the same thing, this lgtm. It may be worth including something similar to those comparisons in the book too, to help users migrate. 

@gakonst Seems this will be a breaking change that may break some scripts/workflows/CIs, so we should try to communicate this once merged since I suspect these are commonly used features

**Comment by gakonst** [comment]
SGTM - Can we document these in the book after we merge so that we can point people to a resource to modify their scripts?

**Comment by jatokz** [comment]
its kind of minimalist,

I can add the regex comparison as well if you think its good

**Comment by gakonst** [comment]
Yeah would love to get more examples in the book as well now that we're on it - go for it. Lgtm pending @mds1 approval

**Comment by mds1** [comment]
I haven’t tested the PR or reviewed the code (prob won’t be able to until after the holidays) but conceptually it lgtm!

**Comment by gakonst** [comment]
@mds1 LMK when you get a chance to try!

**Comment by mds1** [comment]
Testing with `--mt` and seems to be working. Two comments:
1. It only works if I put quotes around the glob, which is tedious. Can we make it work without quotes? For example `forge test --mt test*` finds nothing, but `forge test --mt "test*"` runs all tests
2. Currently it's case sensitive, which I think is consistent with the prior UX right?

**Comment by jatokz** [comment]
1. Yes, I'm going to try to get that fixed soon. Apologies for the caught issues, going to start testing directly aside from unit tests :smile: 

2. I tested prior regex just now and it is consistent

**Comment by jatokz** [comment]
Update on 1) from testing on my machine, since I am using a bash shell the test* was being changed to test because it functioned like a glob in my shell and there was a test directory so it matched to that. One way to verify that this is true for you also is to do `forge test --mt test\*` since that escapes into literal `*`

**Comment by mds1** [comment]
Ah yes escaping with `forge test --mt test\*` does work. But it seems this is true even if there are no directory matches. For example testing on the [forge-std](https://github.com/foundry-rs/forge-std) repo I can't run `forge test --mt *Storage*` to run all storage-related tests. @mattsse @gakonst curious your thoughts on this?

**Comment by gakonst** [comment]
Would like to get this in a place where it doesn't require quotes or escapes, like it is currently..Can we not parse it as String from CLI, and bring it in the right form/global format internally?

**Comment by zerosnacks** [comment]
Is this PR still desired? Personally I prefer the regex pattern as it is much more expressive

Moving this outside of the `1.0` milestone as it is a breaking change

**Comment by mds1** [comment]
I'd be ok closing this, since I haven't seen any complaints about the current behavior

**Pull request closed without merging**